Builds and It's Completely Different but Also Still Build (#18955)
cloudflare/cloudflare-docs: src/content/docs/pages/configuration/git-integration/index.mdx

---
pcx_content_type: concept
title: Git integration
---

> Cloudflare supports connecting Cloudflare Pages to your [GitHub](/pages/configuration/git-integration/github-integration) and [GitLab](/pages/configuration/git-integration/gitlab-integration) repositories, and will automatically deploy your code everytime you push a change to a branch.

Reviewer: ```suggestion
You can connect each Cloudflare Pages project to a [GitHub](/pages/configuration/git-integration/github-integration) or [GitLab](/pages/configuration/git-integration/gitlab-integration) repository, and Cloudflare will automatically deploy your code every time you push a change to a branch.
```

> Using Git integration with Cloudflare also provides:

Reviewer: ```suggestion
When you connect a git repository to your Cloudflare Pages project, Cloudflare will also:
```

> - **Preview deployments for custom branches**, generating preview URLs for a commit to any branch in the repository without affecting your production deployment.

Reviewer: Doesn't seem like this one should be first in the list, since first you kind of have to understand what a preview url is

- **Preview URLs in pull requests** (PRs) to the repository.
- **Build and deployment status checks** within the Git repository.
- **Skipping builds using a commit message**.

These features allow you to manage your deployments directly within GitHub or GitLab without leaving your team's regular development workflow.

> :::caution[You cannot switch to Direct Upload later]

Reviewer: Worth calling out that this is not the case with Workers?

If you deploy using the Git integration, you cannot switch to [Direct Upload](/pages/get-started/direct-upload/) later. However, if you already use a Git-integrated project and do not want to trigger deployments every time you push a commit, you can [disable automatic deployments](/pages/configuration/git-integration/#disable-automatic-deployments) on all branches. Then, you can use Wrangler to deploy directly to your Pages projects and make changes to your Git repository without automatically triggering a build.

:::

## Supported Git Providers

> Cloudflare Pages supports the following Git providers:

Reviewer: It seems simpler to say — we support GitHub and GitLab, except not if you self-host — right?

The pricing/plan level stuff isn't really the relevant detail, it's whether what you're running is cloud hosted or self hosted.

- GitHub

  - [GitHub Free](https://github.com/pricing)
  - [GitHub Team](https://github.com/pricing)
  - [GitHub Enterprise Cloud](https://github.com/pricing)

- GitLab
  - [GitLab Free](https://about.gitlab.com/pricing/)
  - [GitLab Premium](https://about.gitlab.com/pricing/)
  - [GitLab Ultimate](https://about.gitlab.com/pricing/)
  - [GitLab Enterprise](https://about.gitlab.com/enterprise/)

If you use [GitHub Enterprise Server](https://docs.github.com/en/enterprise-server@3.10/admin/overview/about-github-enterprise-server), [Self-Managed GitLab](https://about.gitlab.com/pricing/), or another Git provider (e.g., Bitbucket), you can:

> - Start with a Direct Upload project and deploy using a [CI/CD provider with Wrangler CLI](/pages/how-to/use-direct-upload-with-continuous-integration/).

Reviewer: This seems like the actual recommended path — worth saying `GitHub Actions` by name here since we're talking about GitHub anyways

> - Or, [disable automatic deployments](/pages/configuration/git-integration/#disable-automatic-deployments) in a Git-integrated project and use a [CI/CD provider with Wrangler CLI](/pages/how-to/use-direct-upload-with-continuous-integration/).

Reviewer: This is confusing if I'm starting anything new? Is it necessary to call out in a way here where it's a recommended path? This is a workaround for being restrictive/prescriptive about not mixing together direct upload and pages builds, right?
